gsy_framework/sim_results/carbon_emissions.py

```diff
@@ -0,0 +1,246 @@
+    """
+
+    carbon_emissions = {}
+
+    def __init__(self, api_key: str = None):
+        super().__init__(api_key=api_key)
+
+    def coordinates_to_country_code(self, lat: float, lon: float) -> str:
+        """Get the country code based on the coordinates"""
+        geolocator = Nominatim(user_agent="location_app")
+        location = geolocator.reverse((lat, lon), language="en")
+        if location and "country" in location.raw["address"]:
+            country_code = location.raw["address"]["country_code"].upper()
+            return GEOPY_CONTRY_CODE_TO_ENTSOE_CODE[country_code]
+        return "Country not found"
+
+    def _base_request(
+        self, params: Dict, start: pd.Timestamp, end: pd.Timestamp
+    ) -> requests.Response:
+        """Overwrites EntsoePandasClient _base_request method"""
+        start_str = self._datetime_to_str(start)
+        end_str = self._datetime_to_str(end)
+        base_params = {
+            "securityToken": self.api_key,
+            "periodStart": start_str,
+            "periodEnd": end_str,
+        }
+        params.update(base_params)
+        response = self.session.get(url=ENTSOE_URL, params=params, proxies=None, timeout=None)
+        return response
+
+    def _query_generation_per_plant(
+        self, country_code: Union[Area, str], start: pd.Timestamp, end: pd.Timestamp
+    ):
+        """Overwrites EntsoePandasClient query_generation_per_plant method"""
+        area = lookup_area(country_code)
+        params = {
+            "documentType": "A73",
+            "processType": "A16",
+            "in_Domain": area.code,
+        }
+        response = self._base_request(params=params, start=start, end=end)
+        text = response.text
+        if "Unauthorized" in text:
+            raise ValueError("Invalid API key")
+        df = parse_generation(text, per_plant=True, include_eic=False)
+
+        df.columns = df.columns.set_levels(
+            df.columns.levels[0].str.encode("latin-1").str.decode("utf-8"), level=0
+        )
+        df = df.tz_convert(area.tz)
+        # Truncation will fail if data is not sorted along the index in rare
+        # cases. Ensure the dataframe is sorted:
+        df = df.sort_index(axis=0)
+
+        if df.columns.nlevels == 2:
+            df = (
+                df.assign(newlevel="Actual Aggregated")
+                .set_index("newlevel", append=True)
+                .unstack("newlevel")
+            )
+        df = df.truncate(before=start, after=end)
+        return df
+
+    def calculate_carbon_ratio(
+        self, country_code: str, start: pd.Timestamp, end: pd.Timestamp, stat: str = "median"
+    ) -> pd.DataFrame:
+        """Calculate the total carbon generated based on the specified statistic
+        (min, median, max)"""
+        if stat not in ["min", "median", "max"]:
+            raise ValueError("stat must be one of 'min', 'median', or 'max'")
+
+        df = self._query_generation_per_plant(country_code, start, end)
```

Comment:
Please use better names for these pandas dataframes. Just naming the dataframe as `df` is equal to a normal Python variable being named as `a`

Reply:
Thanks for the roast. Changed aux variables naming